conda/models/prefix_graph.py

```diff
@@ -0,0 +1,387 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from logging import getLogger
+
+from .enums import NoarchType
+from .match_spec import MatchSpec
+from .._vendor.boltons.setutils import IndexedSet
+from ..base.context import context
+from ..common.compat import iteritems, itervalues, odict, on_win
+from ..exceptions import CyclicalDependencyError
+
+log = getLogger(__name__)
+
+
+class PrefixGraph(object):
+    """
+    A directed graph structure used for sorting packages (prefix_records) in prefixes and
+    manipulating packages within prefixes (e.g. removing and pruning).
+
+    The terminology used for edge direction is "parents" and "children" rather than "successors"
+    and "predecessors". The parent nodes of a record are those records in the graph that
+    match the record's "depends" field.  E.g. NodeA depends on NodeB, then NodeA is a child
+    of NodeB, and NodeB is a parent of NodeA.  Nodes can have zero parents, or more than two
+    parents.
+
+    Most public methods mutate the graph.
+    """
+
```

Comment:
The potential inefficiencies I referred to before are

1. Probably some unnecessary calls to toposort
2. ad-hoc creation of `inverted_graph` in several methods